src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientCache.java

```diff
@@ -0,0 +1,179 @@
+   */
+  @VisibleForTesting
+  MongoClientCache(
+      final long keepAliveMS, final long initialCleanUpDelayMS, final long cleanUpDelayMS) {
+    this.keepAliveNanos = TimeUnit.NANOSECONDS.convert(keepAliveMS, TimeUnit.MILLISECONDS);
+    scheduler.scheduleWithFixedDelay(
+        this::checkClientCache, initialCleanUpDelayMS, cleanUpDelayMS, TimeUnit.MILLISECONDS);
+    isAvailable = true;
+  }
+
+  /**
+   * Acquires a new MongoClient
+   *
+   * @param mongoClientFactory the factory to use for creating the MongoClient
+   * @return the MongoClient from the cache or create a new one using the {@code
+   *     MongoClientFactory}.
+   */
+  synchronized MongoClient acquire(final MongoClientFactory mongoClientFactory) {
+    assertIsAvailable();
+    return cacheList.stream()
+        .filter(c -> Objects.equals(c.mongoClientFactory, mongoClientFactory))
+        .findFirst()
+        .orElseGet(
+            () -> {
+              MongoClientCacheItem mongoClientCacheItem =
+                  new MongoClientCacheItem(mongoClientFactory, keepAliveNanos);
+              cacheList.add(mongoClientCacheItem);
+              return mongoClientCacheItem;
+            })
+        .acquire();
+  }
+
+  /**
+   * Releases the MongoClient
+   *
+   * @param mongoClient to release
+   */
+  synchronized void release(final MongoClient mongoClient) {
+    assertIsAvailable();
+    cacheList.stream()
+        .filter(c -> Objects.equals(c.mongoClient, mongoClient))
+        .findFirst()
+        .orElseThrow(() -> new IllegalStateException("MongoClient not in the cache."))
+        .release();
+  }
+
+  void shutdown() {
+    if (isAvailable) {
+      scheduler.shutdown();
+      cacheList.forEach(
+          item -> {
+            try {
+              item.mongoClient.close();
+            } catch (Exception e) {
+              LOGGER.info("Error when shutting down client: {}", e.getMessage());
+            }
+          });
+      cacheList.clear();
+      isAvailable = false;
+    }
+  }
+
+  private synchronized void checkClientCache() {
+    long now = System.nanoTime();
+    cacheList.removeIf(cacheItem -> cacheItem.shouldBeRemoved(now));
+  }
+
+  private void assertIsAvailable() {
+    Assertions.ensureState(
+        () -> isAvailable, "The MongoClientCache has been shutdown and is no longer available");
+  }
+
+  private static final class MongoClientCacheItem {
+    private final MongoClientFactory mongoClientFactory;
+    private final long keepAliveNanos;
+    private final MongoClient mongoClient;
+    private long releaseTime;
+    private int referenceCount;
+
+    private MongoClientCacheItem(
+        final MongoClientFactory mongoClientFactory, final long keepAliveNanos) {
+      this.mongoClientFactory = mongoClientFactory;
+      this.mongoClient = mongoClientFactory.create();
+      this.keepAliveNanos = keepAliveNanos;
+      this.releaseTime = Long.MAX_VALUE;
```

Comment:
Thanks updated.

Reply:
I'm still confused as to why releaseTime being set to MAX_VALUE is not correct.  System.nanoTime cannot be bigger that MAX_VALUE - so its used as a shortcut meaning never release.

I adapted your version, the difference is the keep alive period starts when the MongoClient is fully released. This provides the window for another task to reuse the MongoClient (reducing the costs of creating one).

Reply:
> I adapted your version

I cannot find new commits, probably the latest changes were not pushed.

> System.nanoTime cannot be bigger that MAX_VALUE

This is false, and I think once you see why, the rest of "confused as to why releaseTime being set to MAX_VALUE is not correct" should be clear.

The minimum value of `System.nanoTime` is the [origin](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--), which is unknown an is determined by a JVM process. The maximum value of `System.nanoTime` is origin + 2<sup>63</sup> (may overflow, and that is fine). The total order ≤<sub>nano</sub> is defined via the total order ≤<sub>long</sub> only for values within the interval [origin; origin + 2<sup>63</sup>]: x ≤<sub>nano</sub> y ⇔ x - y ≤<sub>long</sub> 0. Note that the subtraction here is a subtraction in `long`, i.e., it can underflow/overflow.

Now, for example, if the origin is `Long.MAX_VALUE` - 1, then ≤<sub>nano</sub> is defined for `Long.MAX_VALUE` and `Long.MIN_VALUE`, and we can see that `Long.MAX_VALUE` ≤<sub>nano</sub> `Long.MIN_VALUE` because `Long.MAX_VALUE` - `Long.MIN_VALUE` = -1 ≤<sub>long</sub> 0.

In fact, only one value, the origin, is smaller than `Long.MAX_VALUE` in the above example. All the other `System.nanoTime` values, except for the `Long.MAX_VALUE` itself, are bigger than `Long.MAX_VALUE`.

P.S. In [later versions](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.html#nanoTime()) the documentation of the `System.nanoTime` methods was changed from trying to specify the total order to just warning against the pattern `System.nanoTime() >= startTime + timeoutNanos`, which is currently used in the `MongoClientCacheItem` in addition to relying on the assumption that `Long.MAX_VALUE` is the largest `System.nanoTime` value.

Reply:
@stIncMale - apologies for not pushing the changes - I blame being poorly!

Also thank you for the explanation :+1: